### PR TITLE
[PULL REQUEST] Add updates for numerical stability (should be merged before #1170)

### PR DESCRIPTION
**Pull request by yantosca**
## Overview

This PR adds several fixes to improve numerical stability in GEOS-Chem:

  1. In `GeosCore/regrid_a2a_mod.F90`:
     - Use the same values of` PI` and `Re` as in `Headers/physconstants.F90`, which are taken from NIST 2014
     - Change `TINY_R4` to from 1e-20 to 1e-30
     - Now use `IF ( ABS( x ) > 0.0d0 )` instead of `IF ( x == 0.0d0 )`; Due to roundoff, the equality condition may never be met

  2. In `GeosCore/drydep_mod.F90`:
     - Initialize `DRYCOEFF = 0`.  `DRYCOEFF` is used in the HEMCO soil NOx extension, and will have undefined "junk" values if dry deposition is turned off.

  3. In `GeosCore/aerosol_mod.F90`:
     - Do not save TotalOC aerosol diagnostic unless OPOA1 and OPOA2 are defined.  This will avoid a segmentation fault.

  4. In `GeosCore/modis_lai_mod.F90`:
     -  Set the `LANDFRAC` variable to 0 at the start of the loop where it is defined

  5. In `GeosCore/olson_landmap_mod.F90`:
     - Set `State_Met%LandTypeFrac` to zero at the start of the loop where it is defined
     
  6. In `GeosCore/hco_utilities_mod.F90`:
     - Updated comments for clarity

  7. In `KPP/fullchem/fullchem_SulfurChemFuncs.F90`:
     - Reorganize IF blocks to avoid useless computations
     - Replace `State_Het%xRadi(11)` with `State_Chm%AeroRadi(I,J,L,11)`
     - Replace `State_Het%xRadi(12)` with `State_Chm%AeroRadi(I,J,L,12)`
     - Move `USE gckpp_Global` and `USE rateLawUtilFuncs` into the routines where they are referenced

  8. In `GeosCore/fullchem_mod.F90`: 
     - Fix parallelization errors caused by the wrong sequence of routine calls.  See 72c8866ee68a60639be743255bfad2454357af7e for details.

  9. In `GeosCore/hco_interface_gc_mod.F90`:
     - Bug fix: Do not set pointer `Trgt3D => NULL()` in the same line where it is declared.
     - This will make it a global SAVEd variable but we just want it to be a locally-defined automatic variable.

  10. In `GeosCore/tpcore_fvdas_mod.F90`:
      - Make sure that local variables are zeroed before they are used
      - Make sure that arguments declared as `INTENT(OUT)` are zeroed before they are used
      - Removed variables that were not being used in parallel loops
      - Cosmetic changes, updated comments, trimmed trailing whitespace
    
## Related Issues
See: #132, #133, https://github.com/geoschem/geos-chem/issues/1157, https://github.com/geoschem/geos-chem/issues/1103

## Related Pull Requests
  - This PR should be merged before #1166.
  - This PR should be merged concurrently with https://github.com/geoschem/HEMCO/pull/134

### CODE REVIEW

**Review by msulprizio — COMMENTED** [review]
These changes look OK to merge after my two comments for GeosUtil/regrid_a2a_mod.F90 are addressed.

File: GeosUtil/regrid_a2a_mod.F90 [hunk 1]
```diff
@@ -54,22 +54,36 @@ MODULE Regrid_A2A_Mod
...
+  ! GEOS-Chem.  This avoids depending on GEOS-Chem code within the core
+  ! HEMCO modules. (bmy, 7/14/14)
+  !---------------------------------------------------------------------------
+  CHARACTER(LEN=255)    :: NC_DIR        ! Directory w/ netCDF files
+  INTEGER               :: OUTNX         ! # of longitudes (x-dimension) in grid
+  INTEGER               :: OUTNY         ! # of latitudes  (y-dimension) in grid
+  REAL(fp), ALLOCATABLE :: OUTLON (:  )  ! Longitude on output grid
+  REAL(fp), ALLOCATABLE :: OUTSIN (:  )  ! Sines of latitudes on output grid
+  REAL(fp), ALLOCATABLE :: OUTAREA(:,:)  ! Surface areas on output grid
+!
```
[the remark this answers is not in the document]
Comment (msulprizio) [reply GeosUtil/regrid_a2a_mod.F90:71]: I see now that these lines have been copied from the HEMCO version of this module. They do not apply here and should be deleted.

[the remark this answers is not in the document]
Comment (yantosca) [reply GeosUtil/regrid_a2a_mod.F90:71]: Will do

File: GeosUtil/regrid_a2a_mod.F90 [hunk 2]
```diff
@@ -1692,7 +1706,7 @@ SUBROUTINE xmap_r8r8(im, jm, lon1, q1, iin, ilon2, iq2, missval)
   END SUBROUTINE xmap_r8r8
 !EOC
 !------------------------------------------------------------------------------
-!                  GEOS-Chem Global Chemical Transport Model                  !
+!                   Harmonized Emissions Component (HEMCO)                    !
```
[the remark this answers is not in the document]
Comment (yantosca) [reply GeosUtil/regrid_a2a_mod.F90:1709]: Duly noted, will do

File: GeosUtil/regrid_a2a_mod.F90 (hunk 1 above)
[the remark this answers is not in the document]
Comment (yantosca) [reply GeosUtil/regrid_a2a_mod.F90]: Now fixed in commit a635547e3a8bd2a6e4011988c22431fd163eb02a

File: GeosUtil/regrid_a2a_mod.F90 (hunk 2 above)
[the remark this answers is not in the document]
Comment (yantosca) [reply GeosUtil/regrid_a2a_mod.F90]: Now fixed in commit a635547e3a8bd2a6e4011988c22431fd163eb02a

**Comment by yantosca** [comment]
Integration tests pending

**Comment by yantosca** [comment]
Remaining parallelization issues have now been solved in PR #1190.

**Pull request merged**